File: apps/e2e/features/ext/checkin.ts
/**
 * 上课签到的 API 辅助 —— 只做 e2e 的种子数据与读回断言，页面操作留给 spec。
 *
 * 之所以要这一层：投屏上的二维码和口令是服务端每隔 refresh_seconds 秒轮换的，
 * 浏览器里没法「解码二维码」，所以用教师身份读一次 `/live` 拿到当前窗口的
 * token 与口令——这跟学生用手机扫屏幕拿到的是同一个东西。
 */
import { login, req } from '../../core/client'
import { ADMIN_EMAIL, ADMIN_PASSWORD, API_URL, ORG_SLUG } from '../../core/instance'

export interface CheckinSession {
  session_uuid: string
  title: string
  course_uuid: string
  course_name: string
  status: 'open' | 'closed'
  refresh_seconds: number
  code_length: number
  checked_count: number
}

export interface CheckinLive extends CheckinSession {
  token: string
  code: string
  seconds_remaining: number
  recent: Array<{ user_id: number; full_name: string; method: string; ip: string }>
}

function teacherToken(): Promise<string> {
  return login(ADMIN_EMAIL, ADMIN_PASSWORD)
}

/** 开一场签到（教师身份）。 */
export async function createSession(
  courseUuid: string,
  title: string,
  refreshSeconds = 20,
): Promise<CheckinSession> {
  const token = await teacherToken()
  return req<CheckinSession>(
    'POST',
    `/ext/checkin/courses/${courseUuid}/sessions`,
    token,
    { title, refresh_seconds: refreshSeconds },
  )
}

/** 读投屏页那份实时状态：当前 token、口令、剩余秒数、已签人数、最新名单。 */
export async function readLive(sessionUuid: string): Promise<CheckinLive> {
  const token = await teacherToken()
  return req<CheckinLive>('GET', `/ext/checkin/sessions/${sessionUuid}/live`, token)
}

export async function closeSession(sessionUuid: string): Promise<CheckinSession> {
  const token = await teacherToken()
  return req<CheckinSession>('POST', `/ext/checkin/sessions/${sessionUuid}/close`, token)
}

/** 直接调签到接口，返回 HTTP 状态与 detail.code —— 用来断言四种错误码。 */
export async function rawCheck(
  sessionUuid: string,
  body: { token?: string; code?: string },
  bearer: string,
): Promise<{ status: number; code: string | null }> {
  const res = await fetch(
    `${API_URL}/ext/checkin/sessions/${sessionUuid}/check`,
    {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${bearer}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    },
  )
  const text = await res.text()
  let code: string | null = null
  try {
    const parsed = JSON.parse(text)
    code = typeof parsed?.detail?.code === 'string' ? parsed.detail.code : null
  } catch {
    /* 非 JSON 响应，code 保持 null */
  }
  return { status: res.status, code }
}

/**
 * CSV 导出（教师身份）。
 *
 * 返回 `hasBom` 而不是让调用方自己去看正文开头：`Response.text()` 按规范会把
 * 前导 BOM 吃掉，所以只能从原始字节里判断。BOM 是给 Excel 用的，没有它中文会乱码。
 */
export async function exportCsv(
  sessionUuid: string,
): Promise<{ text: string; hasBom: boolean }> {
  const token = await teacherToken()
  const res = await fetch(
    `${API_URL}/ext/checkin/sessions/${sessionUuid}/records?format=csv`,
    { headers: { Authorization: `Bearer ${token}` } },
  )
  if (!res.ok) throw new Error(`CSV export -> ${res.status}`)
  const bytes = new Uint8Array(await res.arrayBuffer())
  const hasBom = bytes[0] === 0xef && bytes[1] === 0xbb && bytes[2] === 0xbf
  return { text: new TextDecoder('utf-8').decode(bytes), hasBom }
}

/**
 * 挑一门课来做签到。
 *
 * `publicOnly` 为真时只挑公开且已发布的——学生端课程页只有这种打得开，
 * 验证课程页顶部的签到入口条要用它。
 */
export async function pickCourse(publicOnly = false): Promise<string> {
  const token = await teacherToken()
  const list = await req<any[]>(
    'GET',
    `/courses/org_slug/${ORG_SLUG}/page/1/limit/100?include_unpublished=true`,
    token,
  )
  const courses = Array.isArray(list) ? list : []
  const found = publicOnly
    ? courses.find((c) => c.public && c.published)
    : courses[0]
  if (!found) {
    throw new Error(
      publicOnly
        ? '本地库里需要至少一门公开且已发布的课程'
        : '本地库里需要至少一门课程',
    )
  }
  return found.course_uuid as string
}

/** 关掉某门课上所有还开着的会话，避免用例之间互相看见对方的入口条。 */
export async function closeAllOpen(courseUuid: string): Promise<void> {
  const token = await teacherToken()
  const sessions = await req<CheckinSession[]>(
    'GET',
    `/ext/checkin/courses/${courseUuid}/sessions`,
    token,
  )
  for (const s of sessions) {
    if (s.status === 'open') await closeSession(s.session_uuid)
  }
}

/** 课程页路径段不带 `course_` 前缀（页面自己会补回去）。 */
export function courseSlug(courseUuid: string): string {
  return courseUuid.replace('course_', '')
}

File: apps/api/src/routers/ext/checkin.py
"""上课签到（SYSU-SAM 扩展）路由。

挂载在 ``/api/v1/ext/checkin``。教师端接口要求对课程有 update 权限，
学生端接口只要求是本组织的登录用户。权限判断全在
``src.services.ext.checkin.checkin`` 里，本文件只做 HTTP 与依赖注入。
"""

from fastapi import APIRouter, Depends, Query, Request, Response
from sqlmodel.ext.asyncio.session import AsyncSession

from src.core.events.database import get_db_session
from src.db.ext.checkin import (
    CheckinLiveRead,
    CheckinPublicSessionRead,
    CheckinRecordRead,
    CheckinSessionCreate,
    CheckinSessionRead,
    CheckinSubmit,
    CheckinSubmitResult,
)
from src.db.users import PublicUser
from src.security.auth import get_current_user
from src.services.ext.checkin.checkin import (
    close_checkin_session,
    create_checkin_session,
    get_live_state,
    get_open_session_for_course,
    get_public_session,
    list_checkin_records,
    list_checkin_sessions,
    records_to_csv,
    submit_checkin,
)

router = APIRouter()


# --------------------------------------------------------------------------
# 教师端
# --------------------------------------------------------------------------


@router.post(
    "/courses/{course_uuid}/sessions",
    response_model=CheckinSessionRead,
    tags=["ext"],
    summary="Start a check-in session for a course",
    description=(
        "Create and immediately open a check-in session. Requires update rights "
        "on the course. The rotating secret is generated server-side and never "
        "returned to any client."
    ),
    responses={
        200: {"description": "Session created and open.", "model": CheckinSessionRead},
        400: {"description": "refresh_seconds out of range"},
        401: {"description": "Authentication required"},
        403: {"description": "User lacks permission on this course"},
        404: {"description": "Course not found"},
    },
)
async def api_create_checkin_session(
    *,
    request: Request,
    course_uuid: str,
    payload: CheckinSessionCreate,
    db_session: AsyncSession = Depends(get_db_session),
    current_user: PublicUser = Depends(get_current_user),
) -> CheckinSessionRead:
    return await create_checkin_session(
        request, db_session, current_user, course_uuid, payload
    )


@router.get(
    "/courses/{course_uuid}/sessions",
    response_model=list[CheckinSessionRead],
    tags=["ext"],
    summary="List check-in sessions of a course",
    description="History of check-in sessions for a course, newest first.",
    responses={
        200: {"description": "Sessions listed."},
        401: {"description": "Authentication required"},
        403: {"description": "User lacks permission on this course"},
        404: {"description": "Course not found"},
    },
)
async def api_list_checkin_sessions(
    *,
    request: Request,
    course_uuid: str,
    db_session: AsyncSession = Depends(get_db_session),
    current_user: PublicUser = Depends(get_current_user),
) -> list[CheckinSessionRead]:
    return await list_checkin_sessions(request, db_session, current_user, course_uuid)


@router.get(
    "/sessions/{session_uuid}/live",
    response_model=CheckinLiveRead,
    tags=["ext"],
    summary="Live state of a check-in session (projector view)",
    description=(
        "Current QR token, numeric code, seconds left in the window, checked-in "
        "count and the 20 most recent records. Polled once per second by the "
        "projector page."
    ),
    responses={
        200: {"description": "Live state.", "model": CheckinLiveRead},
        401: {"description": "Authentication required"},
        403: {"description": "User lacks permission on this course"},
        404: {"description": "Session not found"},
    },
)
async def api_get_checkin_live(
    *,
    request: Request,
    session_uuid: str,
    db_session: AsyncSession = Depends(get_db_session),
    current_user: PublicUser = Depends(get_current_user),
) -> CheckinLiveRead:
    return await get_live_state(request, db_session, current_user, session_uuid)


@router.post(
    "/sessions/{session_uuid}/close",
    response_model=CheckinSessionRead,
    tags=["ext"],
    summary="Close a check-in session",
    description="Closing is idempotent; a closed session rejects further check-ins.",
    responses={
        200: {"description": "Session closed.", "model": CheckinSessionRead},
        401: {"description": "Authentication required"},
        403: {"description": "User lacks permission on this course"},
        404: {"description": "Session not found"},
    },
)
async def api_close_checkin_session(
    *,
    request: Request,
    session_uuid: str,
    db_session: AsyncSession = Depends(get_db_session),
    current_user: PublicUser = Depends(get_current_user),
) -> CheckinSessionRead:
    return await close_checkin_session(request, db_session, current_user, session_uuid)


@router.get(
    "/sessions/{session_uuid}/records",
    tags=["ext"],
    summary="Records of a check-in session",
    description=(
        "Returns JSON by default. Pass ``format=csv`` for a UTF-8-BOM CSV "
        "attachment with name, email, timestamp, method and IP."
    ),
    responses={
        200: {"description": "Records listed."},
        401: {"description": "Authentication required"},
        403: {"description": "User lacks permission on this course"},
        404: {"description": "Session not found"},
    },
)
async def api_list_checkin_records(
    *,
    request: Request,
    session_uuid: str,
    format: str = Query("json", pattern="^(json|csv)$"),
    db_session: AsyncSession = Depends(get_db_session),
    current_user: PublicUser = Depends(get_current_user),
):
    records: list[CheckinRecordRead] = await list_checkin_records(
        request, db_session, current_user, session_uuid
    )
    if format != "csv":
        return records

    live = await get_live_state(request, db_session, current_user, session_uuid)
    csv_body = records_to_csv(records, live.title, live.course_name)
    return Response(
        content=csv_body,
        media_type="text/csv; charset=utf-8",
        headers={
            "Content-Disposition": f'attachment; filename="checkin_{session_uuid}.csv"'
        },
    )


# --------------------------------------------------------------------------
# 学生端
# --------------------------------------------------------------------------


@router.get(
    "/sessions/{session_uuid}/public",
    response_model=CheckinPublicSessionRead,
    tags=["ext"],
    summary="Public view of a check-in session",
    description=(
        "What a student sees after scanning: title, course, status and whether "
        "they already checked in. Never includes the rotating token."
    ),
    responses={
        200: {"description": "Session info.", "model": CheckinPublicSessionRead},
        401: {"description": "Authentication required"},
        403: {"description": "Not a member of this organization"},
        404: {"description": "Session not found"},
    },
)
async def api_get_public_checkin_session(
    *,
    session_uuid: str,
    db_session: AsyncSession = Depends(get_db_session),
    current_user: PublicUser = Depends(get_current_user),
) -> CheckinPublicSessionRead:
    return await get_public_session(db_session, current_user, session_uuid)


@router.post(
    "/sessions/{session_uuid}/check",
    response_model=CheckinSubmitResult,
    tags=["ext"],
    summary="Check in to a session",
    description=(
        "Binds the check-in to the authenticated user — the client cannot pick "
        "who is checked in. Errors carry a machine-readable ``detail.code``: "
        "``already_checked_in`` (409), ``invalid_or_expired_code`` (400), "
        "``session_closed`` (410), ``rate_limited`` (429)."
    ),
    responses={
        200: {"description": "Checked in.", "model": CheckinSubmitResult},
        400: {"description": "Code expired or invalid"},
        401: {"description": "Authentication required"},
        403: {"description": "Not a member of this organization"},
        404: {"description": "Session not found"},
        409: {"description": "Already checked in"},
        410: {"description": "Session closed"},
        429: {"description": "Rate limited"},
    },
)
async def api_submit_checkin(
    *,
    request: Request,
    session_uuid: str,
    payload: CheckinSubmit,
    db_session: AsyncSession = Depends(get_db_session),
    current_user: PublicUser = Depends(get_current_user),
) -> CheckinSubmitResult:
    return await submit_checkin(
        request, db_session, current_user, session_uuid, payload
    )


@router.get(
    "/my/open",
    tags=["ext"],
    summary="Currently open check-in session for a course",
    description=(
        "Used by the student-facing banner on the course and activity pages. "
        "Returns null when the course has no open session."
    ),
    responses={
        200: {"description": "Open session or null."},
        401: {"description": "Authentication required"},
        403: {"description": "Not a member of this organization"},
        404: {"description": "Course not found"},
    },
)
async def api_get_my_open_checkin(
    *,
    course_uuid: str = Query(...),
    db_session: AsyncSession = Depends(get_db_session),
    current_user: PublicUser = Depends(get_current_user),
):
    return await get_open_session_for_course(db_session, current_user, course_uuid)
